### PR TITLE
Don't report `no-process-exit` rule in worker threads

### LINKED ISSUE #328: no-process-exit in worker_threads
[Issuehunt badges](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/issues/328)

Is there a solution for these cases?

image

```js
'use strict'

const {workerData, parentPort} = require('worker_threads')
const debug = require('./debug')
const sleep = require('./sleep')

const {data, seq} = workerData
sleep(3)
  .then(() => {
    // some data processing...
    const res = `modified ${data}`
    parentPort.postMessage(`Data processed - ${seq}: ${res}`)
  })
  .catch(error => {
    debug.error(error)
    process.exit(1)
  })
```

My worker is waiting for an exit event.

```js
'use strict'

const {Worker} = require('worker_threads')

function runService(service, workerData) {
  return new Promise((resolve, reject) => {
    const worker = new Worker(service, {workerData})
    worker.on('message', resolve)
    worker.on('error', reject)
    worker.on('exit', code => {
      if (code !== 0) {
        reject(new Error(`Worker stopped with exit code ${code}`))
      }
    })
  })
}

module.exports = runService
```

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $30.00)

- [issuehunt issuehunt](https://issuehunt.io/u/issuehunt) ($30.00)
### Submitted pull Requests
- [#388 Closes #328: Enable process.exit on thread files](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/pull/388)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
---
IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### PR DESCRIPTION
**Pull request by lubien**
Rule no-process-exit should enable process.exit()
when worker_threads module is required.

Fixes #328

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#328: no-process-exit in worker_threads](https://issuehunt.io/repos/55832243/issues/328)
---

IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### CODE REVIEW

File: test/no-process-exit.js [hunk 1]
```diff
@@ -29,6 +32,9 @@ ruleTester.run('no-process-exit', rule, {
 		'process.once("SIGINT", () => { process.exit(1); })',
 		'process.once("SIGINT", () => process.exit(1))',
 		'process.once("SIGINT", () => { if (true) { process.exit(1); } })',
+		'const {workerData, parentPort} = require(\'worker_threads\')\n\nprocess.exit(1);',
+		'import {workerData, parentPort} from \'worker_threads\'\n\nprocess.exit(1);',
+		'import foo from \'worker_threads\'\n\nprocess.exit(1);',
 		''
```
Comment (fisker) [inline test/no-process-exit.js]: this empty code is introduced in #1 , but seems useleess

File: test/no-process-exit.js [hunk 2]
```diff
@@ -29,6 +32,9 @@ ruleTester.run('no-process-exit', rule, {
 		'process.once("SIGINT", () => { process.exit(1); })',
 		'process.once("SIGINT", () => process.exit(1))',
 		'process.once("SIGINT", () => { if (true) { process.exit(1); } })',
+		'const {workerData, parentPort} = require(\'worker_threads\')\n\nprocess.exit(1);',
```
Comment (fisker) [inline test/no-process-exit.js]:
```suggestion
		outdent`
			const {workerData, parentPort} = require('worker_threads');
			process.exit(1);
		`,
```

File: test/no-process-exit.js [hunk 3]
```diff
@@ -29,6 +32,9 @@ ruleTester.run('no-process-exit', rule, {
...
+		'import {workerData, parentPort} from \'worker_threads\'\n\nprocess.exit(1);',
```
Comment (fisker) [inline test/no-process-exit.js]:
```suggestion
		outdent`
			import {workerData, parentPort} from 'worker_threads';
			process.exit(1);
		`,
```

File: test/no-process-exit.js [hunk 4]
```diff
@@ -29,6 +32,9 @@ ruleTester.run('no-process-exit', rule, {
...
+		'import foo from \'worker_threads\'\n\nprocess.exit(1);',
```
Comment (fisker) [inline test/no-process-exit.js]:
```suggestion
		outdent`
			import foo from 'worker_threads';
			process.exit(1);
		`,
```

File: test/no-process-exit.js [hunk 5]
```diff
@@ -43,6 +49,14 @@ ruleTester.run('no-process-exit', rule, {
 		{
 			code: 'x(process.exit(1));',
 			errors
+		},
+		{
+			code: 'const mod = require(\'not_worker_threads\');\n\nprocess.exit(1);',
```
Comment (fisker) [inline test/no-process-exit.js]:
```suggestion
			code: outdent`
				const mod = require('not_worker_threads');;
				process.exit(1);
			`,
```

File: test/no-process-exit.js [hunk 6]
```diff
@@ -43,6 +49,14 @@ ruleTester.run('no-process-exit', rule, {
...
+			errors
+		},
+		{
+			code: 'import mod from \'not_worker_threads\';\n\nprocess.exit(1);',
```
Comment (fisker) [inline test/no-process-exit.js]:
```suggestion
			code: outdent`
				import mod from 'not_worker_threads';
				process.exit(1);
			`,
```

File: docs/rules/no-process-exit.md [hunk 7]
```diff
@@ -23,3 +23,18 @@ process.on('SIGINT', () => {
     process.exit(1);
 });
 ```
+
+```js
```
Comment (fisker) [inline docs/rules/no-process-exit.md]: Could this example be simpler?

File: test/no-process-exit.js (hunk 1 above)
Comment (lubien) [reply test/no-process-exit.js]: I thought it was meant just to make sure the rule doesn't trigger by nothing. I just left it as is but I can remove it if needed.

File: docs/rules/no-process-exit.md (hunk 7 above)
Comment (lubien) [reply docs/rules/no-process-exit.md]: Yes. I could just import the module, comment the .then case block as "// do stuff". I'm on it

File: test/no-process-exit.js (hunk 1 above)
Comment (lubien) [reply test/no-process-exit.js]: Done

Comment (fisker) [reply test/no-process-exit.js]: Actually, I'm not sure about it, maybe @sindresorhus need review this.

let's keep it unresolved

Comment (lubien) [reply test/no-process-exit.js]: Sure

**Comment by lubien** [comment]
Just realized some commit came from another PR. Removed it and pushed again.

File: test/no-process-exit.js (hunk 1 above)
Comment (sindresorhus) [reply test/no-process-exit.js]: Makes sense to me to keep it. Just to be safe. We should really add it to all the rules.

Comment (lubien) [reply test/no-process-exit.js]: Ok. I'll restore it later today. Should I do it for others here or in a separate PR?

Comment (lubien) [reply test/no-process-exit.js]: Oh it got merged

Comment (sindresorhus) [reply test/no-process-exit.js]: Doesn't matter. We should add it to all the rules.

**Pull request merged**